File: internal/app/server/databases.go
package server

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	"github.com/deeploop-ai/fleet/internal/domain/databases"
	"github.com/deeploop-ai/fleet/internal/domain/projects"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

var identifierRe = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)

type Databases struct {
	projectRepo projects.Repository
	docDB       databases.DocumentDB
}

func NewDatabases(projectRepo projects.Repository, docDB databases.DocumentDB) *Databases {
	return &Databases{projectRepo: projectRepo, docDB: docDB}
}

func (d *Databases) resolveProject(ctx context.Context, projectID string) (*projects.Project, error) {
	p, err := d.projectRepo.GetProject(ctx, projectID)
	if err != nil {
		return nil, err
	}
	if p == nil {
		return nil, status.Error(codes.NotFound, "project not found")
	}
	return p, nil
}

func (d *Databases) CreateDatabase(ctx context.Context, projectID, id, name string) error {
	if name == "" {
		return status.Error(codes.InvalidArgument, "name is required")
	}
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	return d.docDB.CreateDatabase(ctx, projectID, id, name)
}

func (d *Databases) ListDatabases(ctx context.Context, projectID string) ([]databases.Collection, error) {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return nil, err
	}
	return d.docDB.ListDatabases(ctx, projectID)
}

func (d *Databases) GetDatabase(ctx context.Context, projectID, databaseID string) (*databases.Collection, error) {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return nil, err
	}
	return d.docDB.GetDatabase(ctx, projectID, databaseID)
}

func (d *Databases) DeleteDatabase(ctx context.Context, projectID, databaseID string) error {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	return d.docDB.DeleteDatabase(ctx, projectID, databaseID)
}

func (d *Databases) CreateCollection(ctx context.Context, projectID, databaseID, collectionID, name string, attrs []databases.Attribute, idxs []databases.Index, perms []databases.Permission) error {
	if err := d.ValidateIdentifier(databaseID); err != nil {
		return status.Error(codes.InvalidArgument, "database_id is required")
	}
	if err := d.ValidateIdentifier(collectionID); err != nil {
		return status.Error(codes.InvalidArgument, "id is required")
	}
	if name == "" {
		return status.Error(codes.InvalidArgument, "name is required")
	}
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	if len(perms) == 0 {
		perms = databases.DefaultCollectionPermissions()
	}
	return d.docDB.CreateCollection(ctx, projectID, databaseID, collectionID, name, attrs, idxs, perms)
}

func (d *Databases) ListCollections(ctx context.Context, projectID, databaseID string) ([]databases.Collection, error) {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return nil, err
	}
	return d.docDB.ListCollections(ctx, projectID, databaseID)
}

func (d *Databases) GetCollection(ctx context.Context, projectID, databaseID, collectionID string) (*databases.Collection, error) {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return nil, err
	}
	return d.docDB.GetCollection(ctx, projectID, databaseID, collectionID)
}

func (d *Databases) DeleteCollection(ctx context.Context, projectID, databaseID, collectionID string) error {
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	return d.docDB.DeleteCollection(ctx, projectID, databaseID, collectionID)
}

func (d *Databases) CreateAttribute(ctx context.Context, projectID, databaseID, collectionID string, attr databases.Attribute) error {
	if err := d.ValidateIdentifier(databaseID); err != nil {
		return status.Error(codes.InvalidArgument, "database_id is required")
	}
	if err := d.ValidateIdentifier(collectionID); err != nil {
		return status.Error(codes.InvalidArgument, "collection_id is required")
	}
	if err := d.ValidateIdentifier(attr.Key); err != nil {
		return status.Error(codes.InvalidArgument, "key is required")
	}
	if err := d.ValidateAttributeType(attr.Type); err != nil {
		return err
	}
	attr.Type = strings.ToLower(attr.Type)
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	return d.docDB.CreateAttribute(ctx, projectID, databaseID, collectionID, attr)
}

func (d *Databases) CreateIndex(ctx context.Context, projectID, databaseID, collectionID string, idx databases.Index) error {
	if err := d.ValidateIdentifier(databaseID); err != nil {
		return status.Error(codes.InvalidArgument, "database_id is required")
	}
	if err := d.ValidateIdentifier(collectionID); err != nil {
		return status.Error(codes.InvalidArgument, "collection_id is required")
	}
	if err := d.ValidateIndex(idx); err != nil {
		return err
	}
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	return d.docDB.CreateIndex(ctx, projectID, databaseID, collectionID, idx)
}

func (d *Databases) ensureCollection(ctx context.Context, projectID, databaseID, collectionID string, principal databases.Principal) error {
	if err := d.ValidateIdentifier(databaseID); err != nil {
		return status.Error(codes.InvalidArgument, "database_id is required")
	}
	if err := d.ValidateIdentifier(collectionID); err != nil {
		return status.Error(codes.InvalidArgument, "collection_id is required")
	}
	if _, err := d.resolveProject(ctx, projectID); err != nil {
		return err
	}
	col, err := d.docDB.GetCollection(ctx, projectID, databaseID, collectionID)
	if err != nil {
		return err
	}
	if col == nil {
		return status.Error(codes.NotFound, "collection not found")
	}
	return nil
}

func (d *Databases) CreateDocument(
	ctx context.Context,
	projectID, databaseID, collectionID, documentID string,
	data map[string]any,
	perms []databases.Permission,
	principal databases.Principal,
) (*databases.Document, error) {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return nil, status.Error(codes.InvalidArgument, "data is required")
	}
	doc := databases.Document{ID: documentID, Data: data}
	created, err := d.docDB.CreateDocument(ctx, projectID, databaseID, collectionID, doc, perms, principal)
	if err != nil {
		return nil, fmt.Errorf("create document: %w", err)
	}
	got, err := d.docDB.GetDocument(ctx, projectID, databaseID, collectionID, created.ID, principal)
	if err != nil {
		return nil, err
	}
	if got == nil {
		return nil, status.Error(codes.NotFound, "document not found after create")
	}
	return got, nil
}

func (d *Databases) ListDocuments(
	ctx context.Context,
	projectID, databaseID, collectionID string,
	q databases.Query,
	principal databases.Principal,
) ([]databases.Document, int64, string, error) {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return nil, 0, "", err
	}
	list, err := d.docDB.ListDocuments(ctx, projectID, databaseID, collectionID, q, principal)
	if err != nil {
		return nil, 0, "", err
	}
	return list.Documents, list.TotalCount, list.NextPageToken, nil
}

func (d *Databases) GetDocument(
	ctx context.Context,
	projectID, databaseID, collectionID, documentID string,
	principal databases.Principal,
) (*databases.Document, error) {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return nil, err
	}
	doc, err := d.docDB.GetDocument(ctx, projectID, databaseID, collectionID, documentID, principal)
	if err != nil {
		return nil, err
	}
	if doc == nil {
		return nil, status.Error(codes.NotFound, "document not found")
	}
	return doc, nil
}

func (d *Databases) UpdateDocument(
	ctx context.Context,
	projectID, databaseID, collectionID, documentID string,
	data map[string]any,
	perms []databases.Permission,
	principal databases.Principal,
) (*databases.Document, error) {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return nil, err
	}
	if len(data) == 0 && len(perms) == 0 {
		return nil, status.Error(codes.InvalidArgument, "data or permissions is required")
	}
	if len(data) == 0 {
		data = map[string]any{}
	}
	updated, err := d.docDB.UpdateDocument(ctx, projectID, databaseID, collectionID, databases.Document{
		ID:   documentID,
		Data: data,
	}, perms, principal)
	if err != nil {
		return nil, fmt.Errorf("update document: %w", err)
	}
	return &updated, nil
}

func (d *Databases) DeleteDocument(
	ctx context.Context,
	projectID, databaseID, collectionID, documentID string,
	principal databases.Principal,
) error {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return err
	}
	return d.docDB.DeleteDocument(ctx, projectID, databaseID, collectionID, documentID, principal)
}

func (d *Databases) CountDocuments(
	ctx context.Context,
	projectID, databaseID, collectionID string,
	queries []string,
	principal databases.Principal,
) (int64, error) {
	if err := d.ensureCollection(ctx, projectID, databaseID, collectionID, principal); err != nil {
		return 0, err
	}
	return d.docDB.CountDocuments(ctx, projectID, databaseID, collectionID, queries, principal)
}

// MapAttributeType normalizes a validated attribute type to lowercase.
func (d *Databases) MapAttributeType(t string) string {
	return strings.ToLower(t)
}

func (d *Databases) ValidateIdentifier(id string) error {
	if id == "" {
		return status.Error(codes.InvalidArgument, "identifier is required")
	}
	if !identifierRe.MatchString(id) {
		return status.Error(codes.InvalidArgument, fmt.Sprintf("identifier %q must match %s", id, identifierRe.String()))
	}
	return nil
}

func (d *Databases) ValidateAttributeType(t string) error {
	if t == "" {
		return status.Error(codes.InvalidArgument, "type is required")
	}
	switch strings.ToLower(t) {
	case "string", "integer", "float", "boolean", "datetime", "email", "url", "json":
		return nil
	default:
		return status.Error(codes.InvalidArgument,
			fmt.Sprintf("invalid attribute type %q (allowed: string, integer, float, boolean, datetime, email, url, json)", t))
	}
}

func (d *Databases) ValidateIndex(idx databases.Index) error {
	if err := d.ValidateIdentifier(idx.ID); err != nil {
		return status.Error(codes.InvalidArgument, "id is required")
	}
	if idx.Type == "" {
		return status.Error(codes.InvalidArgument, "type is required")
	}
	switch strings.ToLower(idx.Type) {
	case "key", "unique", "fulltext":
	default:
		return status.Error(codes.InvalidArgument,
			fmt.Sprintf("invalid index type %q (allowed: key, unique, fulltext)", idx.Type))
	}
	if len(idx.Attributes) == 0 {
		return status.Error(codes.InvalidArgument, "attributes is required")
	}
	for _, attr := range idx.Attributes {
		if err := d.ValidateIdentifier(attr); err != nil {
			return status.Error(codes.InvalidArgument, fmt.Sprintf("invalid index attribute %q", attr))
		}
	}
	return nil
}

File: sdk/typescript/src/server/databases.ts
import { listQuery, type HttpTransport } from "../http.js";
import type {
  Attribute,
  Collection,
  Database,
  Document,
  Index,
  ListParams,
} from "../types.js";

export class ServerDatabasesService {
  constructor(private readonly http: HttpTransport) {}

  async createDatabase(input: { id: string; name: string }): Promise<Database> {
    return this.http.request<Database>("POST", "/v1/server/databases", {
      auth: "apiKey",
      body: input,
    });
  }

  async listDatabases(params?: ListParams): Promise<Database[]> {
    const res = await this.http.request<{ databases: Database[] }>("GET", "/v1/server/databases", {
      auth: "apiKey",
      query: listQuery(params),
    });
    return res.databases ?? [];
  }

  async getDatabase(id: string): Promise<Database> {
    return this.http.request<Database>("GET", `/v1/server/databases/${id}`, { auth: "apiKey" });
  }

  async deleteDatabase(id: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/server/databases/${id}`, { auth: "apiKey" });
  }

  async createCollection(
    databaseId: string,
    input: {
      id: string;
      name: string;
      permissions?: string[];
      attributes?: Array<{
        key: string;
        type: string;
        size?: number;
        required?: boolean;
        array?: boolean;
      }>;
    }
  ): Promise<Collection> {
    return this.http.request<Collection>(
      "POST",
      `/v1/server/databases/${databaseId}/collections`,
      { auth: "apiKey", body: { database_id: databaseId, ...input } }
    );
  }

  async listCollections(databaseId: string, params?: ListParams): Promise<Collection[]> {
    const res = await this.http.request<{ collections: Collection[] }>(
      "GET",
      `/v1/server/databases/${databaseId}/collections`,
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.collections ?? [];
  }

  async getCollection(databaseId: string, collectionId: string): Promise<Collection> {
    return this.http.request<Collection>(
      "GET",
      `/v1/server/databases/${databaseId}/collections/${collectionId}`,
      { auth: "apiKey" }
    );
  }

  async deleteCollection(databaseId: string, collectionId: string): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/server/databases/${databaseId}/collections/${collectionId}`,
      { auth: "apiKey" }
    );
  }

  async createAttribute(
    databaseId: string,
    collectionId: string,
    input: { key: string; type: string; size?: number; required?: boolean; array?: boolean }
  ): Promise<Attribute> {
    return this.http.request<Attribute>(
      "POST",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/attributes`,
      {
        auth: "apiKey",
        body: { database_id: databaseId, collection_id: collectionId, ...input },
      }
    );
  }

  async createIndex(
    databaseId: string,
    collectionId: string,
    input: { type: string; attributes: string[]; orders?: string[] }
  ): Promise<Index> {
    return this.http.request<Index>(
      "POST",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/indexes`,
      {
        auth: "apiKey",
        body: { database_id: databaseId, collection_id: collectionId, ...input },
      }
    );
  }

  async createDocument(
    databaseId: string,
    collectionId: string,
    input: {
      document_id?: string;
      data: Record<string, unknown>;
      permissions?: string[];
    }
  ): Promise<Document> {
    return this.http.request<Document>(
      "POST",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents`,
      {
        auth: "apiKey",
        body: {
          database_id: databaseId,
          collection_id: collectionId,
          document_id: input.document_id ?? "",
          data: input.data,
          permissions: input.permissions,
        },
      }
    );
  }

  async listDocuments(
    databaseId: string,
    collectionId: string,
    params?: ListParams
  ): Promise<Document[]> {
    const res = await this.http.request<{ documents: Document[] }>(
      "GET",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents`,
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.documents ?? [];
  }

  async getDocument(
    databaseId: string,
    collectionId: string,
    documentId: string
  ): Promise<Document> {
    return this.http.request<Document>(
      "GET",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      { auth: "apiKey" }
    );
  }

  async updateDocument(
    databaseId: string,
    collectionId: string,
    documentId: string,
    data: Record<string, unknown>
  ): Promise<Document> {
    return this.http.request<Document>(
      "PATCH",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      {
        auth: "apiKey",
        body: {
          database_id: databaseId,
          collection_id: collectionId,
          document_id: documentId,
          data,
        },
      }
    );
  }

  async deleteDocument(
    databaseId: string,
    collectionId: string,
    documentId: string
  ): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents/${documentId}`,
      { auth: "apiKey" }
    );
  }

  async countDocuments(
    databaseId: string,
    collectionId: string,
    params?: Pick<ListParams, "queries">
  ): Promise<number> {
    const res = await this.http.request<{ count: number }>(
      "GET",
      `/v1/server/databases/${databaseId}/collections/${collectionId}/documents/count`,
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.count ?? 0;
  }
}
